src/api.js

```
@@ -0,0 +1,146 @@
+
+    while (retries < MAX_RETRIES) {
+        try {
+            if (!token) {
+                const tokenResult = await getToken();
+                token = tokenResult.token;
+                console.log('[AUTH] Token obtained');
+            }
+            console.log('[SHOPS] Fetching for shops.');
+            return await fetchWithRetry(
+                `${API_URL}/v1/coffee_shops`,
+                { method: 'GET' },
+                token,
+                '[SHOPS]',
+            );
+        } catch (err) {
+            if (err.needsNewToken) {
+                token = null;
+                continue;
+            }
+
+            retries++;
+            if (retries >= MAX_RETRIES) {
+                throw new Error(
+                    `[SHOPS] Failed to get coffee shops after ${MAX_RETRIES} attempts: ${err.message}`,
+                );
+            }
+
+            const delay = retries * RETRY_DELAY;
+            console.error(
+                `[SHOPS] Error: ${err.message}. Retrying in ${delay}ms (${retries}/${MAX_RETRIES})`,
+            );
+            await sleep(delay);
+        }
+    }
+}
+
+/**
+ * Get an authentication token from the API
+ * @returns {Promise<{token: string}>} - Object containing the token
+ */
+export async function getToken() {
+    try {
+        return await fetchWithRetry(
+            `${API_URL}/v1/tokens`,
+            { method: 'POST' },
+            null,
+            '[AUTH]',
+        );
+    } catch (err) {
+        throw new Error(`Failed to get authentication token: ${err.message}`);
+    }
+}
+
+/**
+ * Wrapper for fetch with retry logic, timeouts, and token handling
+ * @param {string} url - API endpoint URL
+ * @param {Object} options - Fetch options
+ * @param {string} [token] - Authentication token
+ * @param {string} logPrefix - Prefix for log messages
+ * @returns {Promise<any>} - Parsed JSON response
+ */
+async function fetchWithRetry(url, options = {}, token = null, logPrefix = '') {
```

We noticed that this function attempts to retry the fetch on API errors, which is fair and a good feature for the CLI to have. However `getCoffeeShops` is also doing retries. Seems to us that the responsibility of retrying should be within a single function. 

Can you tell us what happens if `fetchWithRetry` exhausts all it's retries? How would you adapt the code?

---

I've removed the retry logic out of getCoffeeShops. 

At the moment `fetchWithRetry` throws an error when it exhausts all retries and there is no ok response it should throw an error. I now notice that the error thrown outside the while loop is actually redundant as practically it should never be reached.